invenio_sipstore/archivers.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Archivers for SIP."""
+
+from __future__ import absolute_import, print_function
+
+from datetime import datetime
+
+from os import path
+
+from six import BytesIO, b
+
+
+def get_checksum(checksum, expected='md5'):
+    """Return the checksum if the type is the expected."""
+    checksum = checksum.split(':')
+    if checksum[0] != expected or len(checksum) != 2:
+        raise AttributeError('Checksum format not correct.')
+    else:
+        return checksum[1]
+
+
+class BagItArchiver(object):
+    """"BagIt archiver for SIP files."""
+
+    def __init__(self, storage, tags=None):
+        """Constuctor."""
+        self.storage = storage
+        self.tags = tags or {}
+
+    def archive(self, sip, progress_callback):
+        """Archive the SIP generating a BagIt file."""
+        files = sip.sip_files
+        files_info = self.copy_files(files,
+                                     progress_callback=progress_callback)
+
+        self.autogenerate_tags(files_info)
+
+        manifest_file_info = self.create_manifest_file(files_info)
+        bagit_file_info = self.create_bagit_file()
+        baginfo_file_info = self.create_baginfo_file()
+
+        self.create_tagmanifest([manifest_file_info,
+                                 bagit_file_info,
+                                 baginfo_file_info, ])
+
+    def copy_files(self, files, progress_callback):
+        """Copy the files inside the new storage."""
+        result = {}
+
+        total_size = sum([file.file.size for file in files])
+        copied = 0
+
+        old_uri = self.storage.file.uri
+        self.storage.file.uri = path.join(self.storage.file.uri, 'data')
+
+        for file in files:
+            self.storage.filename = file.filepath
```

Comment:
I'm not checking here the possible path traversal attack since this is not the correct way to create a file with a custom name.

How can I manage properly the files inside the storage? If I try to set a path as a filename (i.e. ``my_root/my_docs/test.txt``) it is going to fail due to the storage it is not creating the path if it doesn't exist: https://github.com/inveniosoftware/invenio-files-rest/blob/master/invenio_files_rest/storage.py#L191